From: Nikita Ermakov <arei@altlinux.org>
Date: Thu, 28 Oct 2021 23:31:13 +0300
Subject: loader: drop argv[] argument in grub_initrd_load()

In the case of an error grub_initrd_load() uses argv[] to print the
filename that caused the error. It is also possible to obtain the
filename from the file handles and there is no need to duplicate that
information in argv[], so let's drop it.

Signed-off-by: Nikita Ermakov <arei@altlinux.org>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>

Last-Update: 2022-01-03
Origin: https://lists.gnu.org/archive/html/grub-devel/2021-10/msg00212.html
---
 grub-core/loader/arm/linux.c              | 2 +-
 grub-core/loader/arm64/linux.c            | 2 +-
 grub-core/loader/i386/linux.c             | 2 +-
 grub-core/loader/i386/pc/linux.c          | 2 +-
 grub-core/loader/i386/xen.c               | 3 +--
 grub-core/loader/ia64/efi/linux.c         | 2 +-
 grub-core/loader/linux.c                  | 4 ++--
 grub-core/loader/mips/linux.c             | 2 +-
 grub-core/loader/powerpc/ieee1275/linux.c | 2 +-
 grub-core/loader/sparc64/ieee1275/linux.c | 2 +-
 include/grub/linux.h                      | 2 +-
 11 files changed, 12 insertions(+), 13 deletions(-)

Index: grub2-2.06/grub-core/loader/arm/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/arm/linux.c
+++ grub2-2.06/grub-core/loader/arm/linux.c
@@ -422,7 +422,7 @@ grub_cmd_initrd (grub_command_t cmd __at
   grub_dprintf ("loader", "Loading initrd to 0x%08x\n",
 		(grub_addr_t) initrd_start);
 
-  if (grub_initrd_load (&initrd_ctx, argv, (void *) initrd_start))
+  if (grub_initrd_load (&initrd_ctx, (void *) initrd_start))
     goto fail;
 
   initrd_end = initrd_start + size;
Index: grub2-2.06/grub-core/loader/arm64/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/arm64/linux.c
+++ grub2-2.06/grub-core/loader/arm64/linux.c
@@ -266,7 +266,7 @@ grub_cmd_initrd (grub_command_t cmd __at
       goto fail;
     }
 
-  if (grub_initrd_load (&initrd_ctx, argv, initrd_mem))
+  if (grub_initrd_load (&initrd_ctx, initrd_mem))
     goto fail;
 
   initrd_start = (grub_addr_t) initrd_mem;
Index: grub2-2.06/grub-core/loader/i386/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/i386/linux.c
+++ grub2-2.06/grub-core/loader/i386/linux.c
@@ -1148,7 +1148,7 @@ grub_cmd_initrd (grub_command_t cmd __at
     initrd_mem_target = get_physical_target_address (ch);
   }
 
-  if (grub_initrd_load (&initrd_ctx, argv, initrd_mem))
+  if (grub_initrd_load (&initrd_ctx, initrd_mem))
     goto fail;
 
   grub_dprintf ("linux", "Initrd, addr=0x%x, size=0x%x\n",
Index: grub2-2.06/grub-core/loader/i386/pc/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/i386/pc/linux.c
+++ grub2-2.06/grub-core/loader/i386/pc/linux.c
@@ -462,7 +462,7 @@ grub_cmd_initrd (grub_command_t cmd __at
     initrd_addr = get_physical_target_address (ch);
   }
 
-  if (grub_initrd_load (&initrd_ctx, argv, initrd_chunk))
+  if (grub_initrd_load (&initrd_ctx, initrd_chunk))
     goto fail;
 
   lh->ramdisk_image = initrd_addr;
Index: grub2-2.06/grub-core/loader/i386/xen.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/i386/xen.c
+++ grub2-2.06/grub-core/loader/i386/xen.c
@@ -809,8 +809,7 @@ grub_cmd_initrd (grub_command_t cmd __at
       if (err)
 	goto fail;
 
-      if (grub_initrd_load (&initrd_ctx, argv,
-			    get_virtual_current_address (ch)))
+      if (grub_initrd_load (&initrd_ctx, get_virtual_current_address (ch)))
 	goto fail;
     }
 
Index: grub2-2.06/grub-core/loader/ia64/efi/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/ia64/efi/linux.c
+++ grub2-2.06/grub-core/loader/ia64/efi/linux.c
@@ -563,7 +563,7 @@ grub_cmd_initrd (grub_command_t cmd __at
   grub_dprintf ("linux", "[addr=0x%lx, size=0x%lx]\n",
 		(grub_uint64_t) initrd_mem, initrd_size);
 
-  if (grub_initrd_load (&initrd_ctx, argv, initrd_mem))
+  if (grub_initrd_load (&initrd_ctx, initrd_mem))
     goto fail;
  fail:
   grub_initrd_close (&initrd_ctx);
Index: grub2-2.06/grub-core/loader/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/linux.c
+++ grub2-2.06/grub-core/loader/linux.c
@@ -271,7 +271,7 @@ grub_initrd_close (struct grub_linux_ini
 
 grub_err_t
 grub_initrd_load (struct grub_linux_initrd_context *initrd_ctx,
-		  char *argv[], void *target)
+		  void *target)
 {
   grub_uint8_t *ptr = target;
   int i;
@@ -317,7 +317,7 @@ grub_initrd_load (struct grub_linux_init
 	{
 	  if (!grub_errno)
 	    grub_error (GRUB_ERR_FILE_READ_ERROR, N_("premature end of file %s"),
-			argv[i]);
+			initrd_ctx->components[i].file->name);
 	  grub_initrd_close (initrd_ctx);
 	  return grub_errno;
 	}
Index: grub2-2.06/grub-core/loader/mips/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/mips/linux.c
+++ grub2-2.06/grub-core/loader/mips/linux.c
@@ -452,7 +452,7 @@ grub_cmd_initrd (grub_command_t cmd __at
     initrd_dest = get_physical_target_address (ch) | 0x80000000;
   }
 
-  if (grub_initrd_load (&initrd_ctx, argv, initrd_src))
+  if (grub_initrd_load (&initrd_ctx, initrd_src))
     goto fail;
 
 #ifdef GRUB_MACHINE_MIPS_QEMU_MIPS
Index: grub2-2.06/grub-core/loader/powerpc/ieee1275/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/powerpc/ieee1275/linux.c
+++ grub2-2.06/grub-core/loader/powerpc/ieee1275/linux.c
@@ -363,7 +363,7 @@ grub_cmd_initrd (grub_command_t cmd __at
 
   grub_dprintf ("loader", "Loading initrd at 0x%x, size 0x%x\n", addr, size);
 
-  if (grub_initrd_load (&initrd_ctx, argv, (void *) addr))
+  if (grub_initrd_load (&initrd_ctx, (void *) addr))
     goto fail;
 
   initrd_addr = addr;
Index: grub2-2.06/grub-core/loader/sparc64/ieee1275/linux.c
===================================================================
--- grub2-2.06.orig/grub-core/loader/sparc64/ieee1275/linux.c
+++ grub2-2.06/grub-core/loader/sparc64/ieee1275/linux.c
@@ -413,7 +413,7 @@ grub_cmd_initrd (grub_command_t cmd __at
   grub_dprintf ("loader", "Loading initrd at vaddr 0x%lx, paddr 0x%lx, size 0x%lx\n",
 		addr, paddr, size);
 
-  if (grub_initrd_load (&initrd_ctx, argv, (void *) addr))
+  if (grub_initrd_load (&initrd_ctx, (void *) addr))
     goto fail;
 
   initrd_addr = addr;
Index: grub2-2.06/include/grub/linux.h
===================================================================
--- grub2-2.06.orig/include/grub/linux.h
+++ grub2-2.06/include/grub/linux.h
@@ -21,4 +21,4 @@ grub_initrd_close (struct grub_linux_ini
 
 grub_err_t
 grub_initrd_load (struct grub_linux_initrd_context *initrd_ctx,
-		  char *argv[], void *target);
+		  void *target);
